enterprise/server/cmd/ci_runner/main.go

```diff
@@ -1699,27 +1699,38 @@ func (ws *workspace) fetchRefs(ctx context.Context) error {
 		baseRefs = append(baseRefs, *targetBranch)
 	}
 	// "fork" is referring to the forked repo, if the runner was triggered by a
-	// PR from a fork (forkBranches will be empty otherwise).
-	forkBranches := make([]string, 0)
+	// PR from a fork (forkRefs will be empty otherwise).
+	forkRefs := make([]string, 0)
 
-	// Add the pushed branch to the appropriate list corresponding to the remote
+	// Add the pushed ref to the appropriate list corresponding to the remote
 	// to be fetched (base or fork).
-	isPushedBranchInFork := *targetRepoURL != "" && *pushedRepoURL != *targetRepoURL
-	if isPushedBranchInFork {
-		forkBranches = append(forkBranches, *pushedBranch)
-	} else if *pushedBranch != *targetBranch {
-		baseRefs = append(baseRefs, *pushedBranch)
+	isPushedRefInFork := *targetRepoURL != "" && *pushedRepoURL != *targetRepoURL
+	if isPushedRefInFork {
+		// Try to fetch a branch if possible, because fetching
+		// commits that aren't at the tip of a branch requires
+		// users to set an additional config option (uploadpack.allowAnySHA1InWant)
```

Comment:
If the user force pushes the branch to a different commit before we get to this point, I think the commit could no longer exist and the checkout might fail?

I think we sort of already have this problem today though since we're currently just fetching by branch name so I'd be fine to keep as-is

Reply:
I think it's okay to ignore that case. If the user force pushes a branch, it feels valid that they should not be able to use an older commit that they manually removed from git history

Reply:
sgtm